### ISSUE TITLE
[.NET Core 3][VB.NET][WPF] error BC30149: Class 'Application' must implement 'Sub InitializeComponent()' for interface 'IComponentConnector'.

### ISSUE THREAD

**Issue by carlossanlop** [dotnet/wpf#1971]
_From @onovotny on Tuesday, November 27, 2018 4:08:27 PM_

When building a WPF application using VB.NET on .NET Core 3, you get the following error

`obj\Debug\net462\Application.g.vb(45,16): error BC30149: Class 'Application' must implement 'Sub InitializeComponent()' for interface 'IComponentConnector'.`

This is due to VB.NET applications using `Application.xaml` as the `ApplicationDefinition` item by default, instead of `App.xaml` like C#.

The workaround is to add the following into the `.vbproj` file:

```xml
<ItemGroup>
  <ApplicationDefinition Include="Application.xaml" Condition="'$(EnableDefaultApplicationDefinition)' != 'false' And Exists('$(MSBuildProjectDirectory)/Application.xaml')" />
  <Page Remove="@(ApplicationDefinition)" />
</ItemGroup>
```

_Copied from original issue: dotnet/core#2080_

**Comment by carlossanlop** [comment]
_From @karelz on Tuesday, November 27, 2018 6:13:13 PM_

cc @diverdan92 @fabiant3 @vatsan-madhavan

**Comment by carlossanlop** [comment]
_From @vatsan-madhavan on Tuesday, November 27, 2018 6:19:03 PM_

This is related to https://github.com/dotnet/cli/issues/10106. We should consider updating the default logic in the SDK for VB.NET to incorporate Application.xaml as the default for Preview 2. 

cc @dotnet/wpf-developers

**Comment by carlossanlop** [comment]
_From @nguerrera on Tuesday, November 27, 2018 7:38:32 PM_

I expected the following to work, but it does not:

``` xml
<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <OutputType>WinExe</OutputType>
    <UseWPF>true</UseWPF>
    <!-- Don't make ApplicationDefinition for me -->
    <EnableDefaultApplicationDefinition>false</EnableDefaultApplicationDefinition>
  </PropertyGroup>

  <ItemGroup>
    <!-- This is my application definition -- I should be able to call it whatever I want -->
    <ApplicationDefinition Include="Application.xaml" />
  </ItemGroup>
</Project>
```

So there are two issues here:

1. Default application definition for VB doesn't match VB tradition
2. If you opt out of EnableApplicationDefinition, but not EnabledDefaultPageItems, you have to remove your own ApplicationDefinition from pages yourself.

**Comment by carlossanlop** [comment]
_From @nguerrera on Tuesday, November 27, 2018 7:51:23 PM_

I think this is a simpler workaround if you're going to paste it into your project and not a targets file trying to light-up on Application.xaml  presence and different values of EnableDefaultApplicationDefinition.

``` diff
<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <OutputType>WinExe</OutputType>
    <UseWPF>true</UseWPF>
+    <EnableDefaultApplicationDefinition>false</EnableDefaultApplicationDefinition>
  </PropertyGroup>

  <ItemGroup>
+    <ApplicationDefinition Include="Application.xaml" />
+    <Page Remove="@(ApplicationDefinition)" />
  </ItemGroup>
</Project>
```

It is basically what I expected to work + the Page remove as a workaround for bug (2).

**Comment by carlossanlop** [comment]
_From @nguerrera on Tuesday, November 27, 2018 7:54:43 PM_

I wonder if it's worth making the targets have different defaults for VB or C#. Another solution would be for default to try App.xaml first and fall back Application.xaml (based on existence), irrespective of language. That feels less surprising to me. If I were porting something from VB to C# or vice versa, I'd get very confused otherwise.

**Comment by carlossanlop** [comment]
_From @vatsan-madhavan on Tuesday, November 27, 2018 8:19:10 PM_

I like the idea of treating both App.xaml and Application.xaml as default candidates for ApplicationDefinition.

**Comment by carlossanlop** [comment]
_From @DustinCampbell on Wednesday, November 28, 2018 9:35:16 PM_

I agree. Supporting both `App.xaml` and `Application.xaml` for both languages seems sensible.

**Comment by carlossanlop** [comment]
_From @rrvenki on Friday, June 14, 2019 9:32:23 AM_

1. In App.xaml, remove the StartupURI element in <Application> tag.
2. Change App.xaml.vb as below.

```
Imports System.Windows
''' <summary>
''' Interaction logic for App.xaml
''' </summary>
Partial Public Class App
    'Inherits Application
    Public Shared Sub Main()
        Dim app As Application = New Application()
        app.Run(New MainWindow())
    End Sub
End Class
```
3. Add "StartupObject" in the vbproj file as below.
```
<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">

  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <UseWPF>true</UseWPF>
    <StartupObject>Sub Main</StartupObject>
  </PropertyGroup>
</Project>
```

Error BC30149 will go off and the project will compile fine.

**Comment by Berrysoft** [comment]
It should be fixed by #545, but the fix might be removed because Visual Basic is not ready for WPF...

**Comment by vatsan-madhavan** [comment]
Yes, we fixed this via #545. The decision IIRC was to retain the conventions as before -  Application.xaml for VB and App.xaml for C#.
